worker/codalabworker/local_run/docker_image_manager.py

```diff
@@ -0,0 +1,183 @@
+from collections import namedtuple
+import threading
+import os
+import time
+import traceback
+import logging
+import json
+
+from codalabworker.formatting import size_str
+from codalabworker.docker_client import DockerException
+from codalabworker.fsm import (
+    BaseDependencyManager,
+    JsonStateCommitter,
+    DependencyStage,
+    StateTransitioner
+)
+
+logger = logging.getLogger(__name__)
+
+DockerImageState = namedtuple('DockerImageState', 'stage digest killed last_used message')
+
+class DockerImageManager(StateTransitioner, BaseDependencyManager):
+
+    def __init__(self, docker, state_committer, max_images_bytes, max_age_failed_seconds=60):
+        super(DockerImageManager, self).__init__()
+        self.add_transition(DependencyStage.DOWNLOADING, self._transition_from_DOWNLOADING)
+        self.add_transition(DependencyStage.READY, self._transition_from_READY)
+        self.add_transition(DependencyStage.FAILED, self._transition_from_FAILED)
+
+        self._state_committer = state_committer
+        self._docker = docker
+        self._images = {} # digest -> DockerImageState
+        self._downloading = {}
```

Comment:
comment

Reply:
This isn't serialized - is there a problem there?

Reply:
this dict is for keeping track of the download threads so no need to serialize.